neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsListIterator.java

```diff
@@ -0,0 +1,594 @@
+     */
+    @SuppressWarnings("unchecked")
+    private void setValues(IterationHelper helper, Object value) {
+        List<E> valuesList = List.class.cast(value);
+        final int firstIndex = helper.getFirstIndex(valuesList);
+
+        if (resolve()) {
+            values = valuesList.listIterator(firstIndex);
+        }
+        else {
+            values = InternalEList.class.cast(valuesList).basicListIterator(firstIndex);
+        }
+    }
+
+    /**
+     * Resets the iterator on values for the last prepared multi-valued feature.
+     */
+    private void resetValues() {
+        values = null;
+    }
+
+    /**
+     * Returns {@code true} if the iterator on values for the last prepared multi-valued feature has more elements when
+     * traversing the list in the defined direction.
+     *
+     * @param helper the iteration helper that defines the direction to use
+     *
+     * @return {@code true} if the iterator on values for the last prepared multi-valued feature has more elements
+     */
+    private boolean hasRemainingValues(IterationHelper helper) {
+        checkState(nonNull(values), "The current feature is neither multi-valued or a feature map");
+
+        // Not a feature map
+        if (!isFeatureMap) {
+            return helper.hasMoreElements(values);
+        }
+
+        // Filter the entries of the feature map
+        while (helper.hasMoreElements(values)) {
+            E value = helper.advance(values);
+            FeatureMap.Entry entry = FeatureMap.Entry.class.cast(value);
+            if (isIncludedEntry(entry)) {
+                helper.reverse(values);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Advances the iterator on values for the current multi-valued feature, and returns the value associated to the
+     * specified {@code feature}.
+     *
+     * @param helper  the iteration helper that defines the direction to use
+     * @param feature the multi-valued feature associated with the value
+     *
+     * @return the result of the iteration
+     */
+    @Nonnull
+    private FutureResult<E> getValue(IterationHelper helper, EStructuralFeature feature) {
+        checkState(nonNull(values), "The current feature is neither multi-valued or a feature map");
+
+        final Object value = helper.advance(values);
+        final IterationDirection direction = helper.direction();
+
+        return isFeatureMap
+                ? new FutureResult<>(direction, FeatureMap.Entry.class.cast(value))
+                : new FutureResult<>(direction, value, feature);
+    }
+
+    // endregion
+
+    /**
+     * A prepared result.
+     *
+     * @param <E> the type of the value
+     */
+    @ParametersAreNonnullByDefault
+    private static class FutureResult<E> {
+
+        /**
+         * An undefined result.
+         */
+        @Nonnull
+        private static final FutureResult<?> UNDEFINED = new Empty<>();
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove this reference to "Empty". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2390)